client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.tsx

```diff
@@ -72,29 +87,64 @@ const ProfessionalEmailUpsell = ( {
 
 	const optionalMailboxFields = [ 'alternativeEmail', 'name' ];
 
+	const changeIntervalLength = ( newIntervalLength: IntervalLength ) => {
+		setSelectedIntervalLength( newIntervalLength );
+		dispatch(
+			recordTracksEvent( 'calypso_professional_email_upsell_nudge_billing_interval_toggle_click', {
+				domain_name: domainName,
+				new_interval: newIntervalLength,
+			} )
+		);
+	};
+
+	/**
+	 * We calculate the price for a year subscription, given how many months we are going to offer for free.
+	 * It takes into account leap years, and months like february.
+	 *
+	 * Example: If we give 3 months for free in February, for a product that cost 35$ yearly then we return the price
+	 * that we need to bill after those 3 months for the rest of the year.
+	 *
+	 * @param productCost
+	 * @param freeMonths
+	 */
+	const getProratedPrice = ( productCost: number, freeMonths: number ) => {
+		const now = new Date();
+		const freeDays = new Date( new Date().setMonth( now.getMonth() + freeMonths ) ).getTime();
```

Comment:
There is a small logic bug on this line. The core issue is that our billing system does rollovers differently than the JS date functions. If we have a subscription that starts on the 31st of January, it will renew on the last day of each month, or, if we're dealing with a three month trial, the trial will end on the 30th of April. However, the `setMonth()` function rolls over to the next month if the new month has fewer days than the current date.
If you apply this logic starting on 31 January, we will be off by 1 day, because the trial expiry date will be computed as 1 May, not 30 April.